extension/storage/local/filestorage/client.go

```diff
@@ -0,0 +1,96 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filestorage
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.etcd.io/bbolt"
+)
+
+var defaultBucket = []byte(`default`)
+
+type fileStorageClient struct {
+	db *bbolt.DB
+}
+
+func newClient(file string, timeout time.Duration) (*fileStorageClient, error) {
+	options := &bbolt.Options{Timeout: timeout}
+	db, err := bbolt.Open(file, 0666, options)
+	if err != nil {
+		return nil, err
+	}
+
+	initBucket := func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(defaultBucket)
+		return err
+	}
+	if err := db.Update(initBucket); err != nil {
+		return nil, err
+	}
+
+	return &fileStorageClient{db}, nil
+}
+
+// Get will retrieve data from storage that corresponds to the specified key
+func (c *fileStorageClient) Get(_ context.Context, key string) ([]byte, error) {
+	var result []byte
+	get := func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(defaultBucket)
+		if bucket == nil {
+			return fmt.Errorf("storage not initialized")
+		}
+		result = bucket.Get([]byte(key))
+		return nil // no error
+	}
+
+	if err := c.db.Update(get); err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// Set will store data. The data can be retrieved using the same key
+func (c *fileStorageClient) Set(_ context.Context, key string, value []byte) error {
+	set := func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(defaultBucket)
+		if bucket == nil {
+			return fmt.Errorf("storage not initialized")
+		}
+		return bucket.Put([]byte(key), value)
+	}
+
+	return c.db.Update(set)
+}
+
+// Delete will delete data associated with the specified key
```

Comment:
Does this return an error if the key does not exist?

Reply:
The short answer is no.

The longer version: I think that all methods on this interface should only return an error if a "real" problem occurred (such as the storage not being available). 

There are a couple reasons why it makes more sense this way, but of course am open to feedback:
1. I think we need a way to differentiate between data operations [overwrite | not-found | no-op] from "real" problems.
2. Since we're working on a key/value store in Go, I think it makes sense to mirror the behavior of a Golang `map`. This means `Set` doesn't error if key already exists - it just overwrites the value. `Get` doesn't error if a key is not found - it just returns `nil`. `Delete` doesn't error if the key doesn't exist - it just no-ops.

That said, I think there is a use case for returning an indication of whether or not anything was deleted. However, one could call `Get` immediately before `Delete` if this needs to be ascertained.

Reply:
I agree with all you said. Please document in the code.